File: src/test/java/de/zalando/zally/rules/NoVersionInUriRuleTest.java
package de.zalando.zally.rules;

import de.zalando.zally.Violation;
import de.zalando.zally.ViolationType;
import io.swagger.models.Swagger;
import org.junit.Test;

import java.util.List;

import static org.junit.Assert.assertEquals;

public class NoVersionInUriRuleTest {
    private final NoVersionInUriRule rule = new NoVersionInUriRule();

    private void assertViolationOccurs(String basePath) {
        Swagger swagger = new Swagger();
        swagger.basePath(basePath);

        List<Violation> violations = rule.validate(swagger);
        assertEquals(1, violations.size());

        Violation violation = violations.get(0);
        assertEquals("Do Not Use URI Versioning", violation.getTitle());
        assertEquals("basePath attribute contains version number", violation.getDescription());
        assertEquals(ViolationType.MUST, violation.getViolationType());
        assertEquals(rule.LINK, violation.getRuleLink());
    }

    @Test
    public void returnsViolationsWhenVersionIsInTheBeginingOfBasePath() {
        assertViolationOccurs("/v1/tests");
    }

    @Test
    public void returnsViolationsWhenVersionIsInTheMiddleOfBasePath() {
        assertViolationOccurs("/api/v1/tests");
    }

    @Test
    public void returnsViolationsWhenVersionIsInTheEndOfBasePath() {
        assertViolationOccurs("/api/v1");
    }

    @Test
    public void returnsViolationsWhenVersionIsBig() {
        assertViolationOccurs("/v1024/tests");
    }

    @Test
    public void returnsEmptyViolationListWhenNoVersionFoundInURL() {
        Swagger swagger = new Swagger();
        swagger.basePath("/violations/");

        List<Violation> violations = rule.validate(swagger);
        assertEquals(0, violations.size());
    }
}

File: server/src/test/java/de/zalando/zally/rules/NoVersionInUriRuleTest.kt
package de.zalando.zally.rules

import de.zalando.zally.Violation
import de.zalando.zally.ViolationType
import io.swagger.models.Swagger
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class NoVersionInUriRuleTest {

    val expectedViolation = Violation(
            NoVersionInUriRule(),
            "Do Not Use URI Versioning",
            "basePath attribute contains version number",
            ViolationType.MUST,
            NoVersionInUriRule().url,
            emptyList())

    @Test
    fun returnsViolationsWhenVersionIsInTheBeginingOfBasePath() {
        val swagger = Swagger().apply { basePath = "/v1/tests" }
        assertThat(NoVersionInUriRule().validate(swagger)).isEqualTo(expectedViolation)
    }

    @Test
    fun returnsViolationsWhenVersionIsInTheMiddleOfBasePath() {
        val swagger = Swagger().apply { basePath = "/api/v1/tests" }
        assertThat(NoVersionInUriRule().validate(swagger)).isEqualTo(expectedViolation)
    }

    @Test
    fun returnsViolationsWhenVersionIsInTheEndOfBasePath() {
        val swagger = Swagger().apply { basePath = "/api/v1" }
        assertThat(NoVersionInUriRule().validate(swagger)).isEqualTo(expectedViolation)
    }

    @Test
    fun returnsViolationsWhenVersionIsBig() {
        val swagger = Swagger().apply { basePath = "/v1024/tests" }
        assertThat(NoVersionInUriRule().validate(swagger)).isEqualTo(expectedViolation)
    }

    @Test
    fun returnsEmptyViolationListWhenNoVersionFoundInURL() {
        val swagger = Swagger().apply { basePath = "/violations/" }
        assertThat(NoVersionInUriRule().validate(swagger)).isNull()
    }

    @Test
    fun returnsEmptyViolationListWhenBasePathIsNull() {
        val swagger = Swagger()
        assertThat(NoVersionInUriRule().validate(swagger)).isNull()
    }
}
